docs/proposals/cluster-autoscaling.md

```diff
@@ -0,0 +1,314 @@
+  Google         |                         |                     |
+                 |                         |                     |
+                 |                  +------v-----+       +-------+------+
+                 |                  |Google Cloud|       |MIG Definition|
+                 |                  | Monitoring |       +-------^------+
+                 |                  +------------+               |
+                 |                         |                     |
+                 |                         |                     |
+                 |                  +------v-----+               |
+                 +----------------->|Google Cloud+---------------+
+                                    | Autoscaler |
+                                    +------------+
+```
+
+
+We will also configure GCE Autoscaler so that it deletes nodes  in a non-random way. There are couple strategies for killing to be considered:
+
+* Node with the fewest number of pods.
+* Node with the fewest number of non-replicated pods.
+* Node with the lowest load.
+
+GCA will know about our preferences from the metrics that would go to GCM the similar way as cpu-usage (it will kill nodes starting from the one with the lowest metric). As for 1.2 the corresponding API may not be ready we will use a predefined metric name and suffix for GCA name.
+
+We will use MIG shutdown script to support pod graceful terminate in case of node deletion.
+
+This will solve the following problems:
+
+* (I1) [GCE specific API] we will have a single tool for all autoscalers (kubectl) with defined configuration
+* (I4) [Autoscaler kills nodes at random] We will use simplest metric (number of pods per node); based on user
+feedback we may enhance it and include more advanced logic by computing per node metrics inside KAC (to take
+into account what pods are running on each node). It will also help greatly with (I7).
+* (I5) [Hard to debug] KAC will generate events that correspond to GCA decisions based on gcloud compute operations list.
+* (I6) [Hard to enable/disable] user can easily update the configuration on a running cluster.
+
+It will also allow us to switch the implementation later without changing the interface (too much).
+
+This will not solve the following problems:
+
+(I2) [GCE specific logic] To port this solution to other cloud providers one would have to reimplement whole logic that
+implements common interface (or configure cloud-provider specific autoscaler); it’s possible but hard.
+(I3) [Works only for well formed pods] We may also introduce an experimental/hacky metrics for pending pods that will be useful
+for scaling up clusters when the pods are larger than 1-target. This will be only a help, but will not be a full solution.
+
+No matter where the cluster autoscaling config will be stored it will need a definition. For development/experimentation/proof-of-concept it
+will be stored in APIServer.
```

Comment:
This is the second time you've alluded to storing it somewhere other than API server. Where else would you store it? I thought all of the config proposals still involve storing the data in the API server. (And actually, to be precise, no data is stored in the API server -- it's stored in etcd.)